File: scripts/wasm/serve.mjs
// Minimal static file server for the wasm app build (no dependencies; runs on
// the emsdk-bundled node via serve-app.ps1). Serves correct MIME types for
// .wasm/.js/.data, which generic file servers sometimes get wrong.
// Usage: node serve.mjs <root-dir> [port] [--open]
import { createServer } from 'node:http';
import { spawn } from 'node:child_process';
import { readFile } from 'node:fs/promises';
import { extname, join, normalize, sep } from 'node:path';

// Flags are filtered out before the positional arguments so --open can sit
// anywhere on the command line.
const args = process.argv.slice(2);
const openInDefaultBrowser = args.includes('--open');
const positional = args.filter((argument) => !argument.startsWith('--'));
const root = normalize(positional[0] ?? '.');
const port = Number(positional[1] ?? 8973);
const types = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript',
  '.mjs': 'text/javascript',
  '.wasm': 'application/wasm',
  '.data': 'application/octet-stream',
  '.json': 'application/json',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.ico': 'image/x-icon',
};

createServer(async (request, response) => {
  try {
    const url = new URL(request.url, 'http://localhost');
    let pathname = decodeURIComponent(url.pathname);
    if (pathname.endsWith('/')) {
      pathname += 'patchy.html';
    }
    const file = normalize(join(root, pathname));
    if (file !== root && !file.startsWith(root + sep)) {
      throw new Error('outside root');
    }
    // Serve a precompressed variant when the client accepts it and staging
    // produced one (scripts/wasm/precompress-site.mjs), mirroring the
    // deployed .htaccess rewrite so the staged-site check exercises the same
    // encoded responses as production.
    const acceptEncoding = String(request.headers['accept-encoding'] ?? '');
    let body;
    let contentEncoding;
    for (const [token, suffix] of [['br', '.br'], ['gzip', '.gz']]) {
      if (body === undefined && acceptEncoding.includes(token)) {
        try {
          body = await readFile(file + suffix);
          contentEncoding = token;
        } catch {
          // No precompressed variant; fall through to the identity file.
        }
      }
    }
    body ??= await readFile(file);
    response.writeHead(200, {
      'Content-Type': types[extname(file).toLowerCase()] ?? 'application/octet-stream',
      ...(contentEncoding ? { 'Content-Encoding': contentEncoding, 'Vary': 'Accept-Encoding' } : {}),
      'Cache-Control': 'no-store',
      // Cross-origin isolation, matching the deployed .htaccess: the
      // multithreaded app needs SharedArrayBuffer, which browsers only enable
      // when these two headers are present. Keeping them here makes the local
      // dev loop match production.
      'Cross-Origin-Opener-Policy': 'same-origin',
      'Cross-Origin-Embedder-Policy': 'require-corp',
    });
    response.end(body);
  } catch {
    response.writeHead(404);
    response.end('not found');
  }
}).listen(port, '127.0.0.1', () => {
  const url = `http://localhost:${port}/`;
  console.log(`serving ${root} at ${url}patchy.html`);
  // Opening from the listen callback is the only race-free moment: the socket
  // is accepting before the browser is told to load anything.
  if (openInDefaultBrowser) {
    open_url(url);
  }
});

// Hands the URL to the OS default browser. Detached and stdio-ignored so
// Ctrl+C on the server never reaches the browser and a chatty opener cannot
// block; the error handler matters because a missing opener (xdg-open on a
// bare container) arrives as an async 'error' event that would otherwise take
// the server down with it.
function open_url(url) {
  const [command, commandArgs] =
    process.platform === 'win32'
      ? ['cmd', ['/c', 'start', '', url]]
      : process.platform === 'darwin'
        ? ['open', [url]]
        : ['xdg-open', [url]];
  const child = spawn(command, commandArgs, { detached: true, stdio: 'ignore' });
  child.on('error', () => {
    console.log(`could not open a browser automatically; open ${url} yourself`);
  });
  child.unref();
}

File: scripts/wasm/serve.py
#!/usr/bin/env python3
"""Static file server for the wasm app build (stdlib only, python 3.9+).

Mirror of serve.mjs for machines without node (studiomac): sends the COOP/COEP
headers cross-origin isolation needs, the wasm MIME types, and the .br/.gz
precompressed-variant negotiation. Additionally accepts POST /memtest-sample
from scripts/wasm/stress-harness.html and appends one JSON line per sample to
the results directory, so a memory run needs no channel out of the browser
beyond same-origin fetch.

Usage: python3 serve.py <root-dir> [port] [--results-dir <dir>]
Binds 127.0.0.1 only. Exits non-zero if the port is already taken.
"""

import http.server
import json
import re
import sys
import threading
from pathlib import Path

TYPES = {
    ".html": "text/html; charset=utf-8",
    ".js": "text/javascript",
    ".mjs": "text/javascript",
    ".wasm": "application/wasm",
    ".data": "application/octet-stream",
    ".json": "application/json",
    ".svg": "image/svg+xml",
    ".png": "image/png",
    ".ico": "image/x-icon",
}

RUN_ID_RE = re.compile(r"^[A-Za-z0-9_.-]{1,64}$")


class Handler(http.server.BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"
    root = Path(".")
    results_dir = None
    telemetry_lock = threading.Lock()

    def do_GET(self):  # noqa: N802 (BaseHTTPRequestHandler contract)
        try:
            path = self.path.split("?", 1)[0].split("#", 1)[0]
            if path.endswith("/"):
                path += "patchy.html"
            file = (self.root / path.lstrip("/")).resolve()
            if self.root != file and self.root not in file.parents:
                raise FileNotFoundError(path)
            body = None
            content_encoding = None
            accept = self.headers.get("Accept-Encoding", "")
            for token, suffix in (("br", ".br"), ("gzip", ".gz")):
                variant = file.with_name(file.name + suffix)
                if body is None and token in accept and variant.is_file():
                    body = variant.read_bytes()
                    content_encoding = token
            if body is None:
                body = file.read_bytes()
            self.send_response(200)
            self.send_header(
                "Content-Type", TYPES.get(file.suffix.lower(), "application/octet-stream")
            )
            if content_encoding:
                self.send_header("Content-Encoding", content_encoding)
                self.send_header("Vary", "Accept-Encoding")
            self.send_header("Content-Length", str(len(body)))
            self.send_header("Cache-Control", "no-store")
            # Cross-origin isolation, matching the deployed .htaccess: the
            # multithreaded app needs SharedArrayBuffer, which browsers only
            # enable when these two headers are present.
            self.send_header("Cross-Origin-Opener-Policy", "same-origin")
            self.send_header("Cross-Origin-Embedder-Policy", "require-corp")
            self.end_headers()
            self.wfile.write(body)
        except (OSError, ValueError):
            self.send_error(404, "not found")

    def do_POST(self):  # noqa: N802
        # /memtest-sample, not /telemetry: "telemetry" URLs sit on standard
        # content-blocker lists (EasyPrivacy and friends) and the POSTs die as
        # ERR_BLOCKED_BY_CLIENT, silently gutting open-driver kill detection.
        if self.path.split("?", 1)[0] != "/memtest-sample" or self.results_dir is None:
            self.send_error(404, "not found")
            return
        try:
            length = int(self.headers.get("Content-Length", "0"))
            raw = self.rfile.read(min(length, 1 << 20))
            sample = json.loads(raw)
            run_id = str(sample.get("runId", "default"))
            if not RUN_ID_RE.match(run_id):
                run_id = "default"
            line = json.dumps(sample, separators=(",", ":"))
            with self.telemetry_lock:
                self.results_dir.mkdir(parents=True, exist_ok=True)
                out = self.results_dir / f"telemetry-{run_id}.jsonl"
                with out.open("a", encoding="utf-8") as handle:
                    handle.write(line + "\n")
            self.send_response(204)
            self.send_header("Content-Length", "0")
            self.end_headers()
        except (OSError, ValueError):
            self.send_error(400, "bad telemetry")

    def log_message(self, fmt, *args):
        # Page GETs are the open-driver kill signal (a memory-killed tab
        # auto-reloads), so log html requests only; asset noise is dropped.
        if ".html" in self.path or self.path == "/":
            sys.stderr.write("%s - %s\n" % (self.log_date_time_string(), fmt % args))


def main():
    args = [a for a in sys.argv[1:] if not a.startswith("--")]
    root = Path(args[0] if args else ".").resolve()
    port = int(args[1]) if len(args) > 1 else 8973
    results = None
    if "--results-dir" in sys.argv:
        results = Path(sys.argv[sys.argv.index("--results-dir") + 1]).resolve()
    if not root.is_dir():
        sys.exit(f"serve.py: root directory not found: {root}")
    Handler.root = root
    Handler.results_dir = results
    try:
        server = http.server.ThreadingHTTPServer(("127.0.0.1", port), Handler)
    except OSError as error:
        sys.exit(f"serve.py: cannot bind 127.0.0.1:{port} ({error})")
    print(f"serving {root} at http://localhost:{port}/patchy.html", flush=True)
    if results:
        print(f"telemetry to {results}", flush=True)
    server.serve_forever()


if __name__ == "__main__":
    main()
